File: leetcode/238.除自身以外数组的乘积/238.除自身以外数组的乘积.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        b, tmp = [1] * n, 1
        for i in range(1, n):
            b[i] = b[i - 1] * nums[i - 1]
        for i in range(n - 2, -1, -1):
            tmp *= nums[i + 1]
            b[i] *= tmp
        return b


if __name__ == '__main__':
    print(Solution().productExceptSelf([1, 2, 3, 4]))

File: leetcode/238.除自身以外数组的乘积/238.除自身以外数组的乘积.go
package main

func productExceptSelf(nums []int) []int {
	// 计算乘积时跳过 nums[1]
	// 前缀乘积*后缀乘积
	n := len(nums)
	res := make([]int, n)
	for i := 0; i < n; i++ {
		res[i] = 1
	}
	// 前缀乘积，下三角，跳过nums[0]
	for i := 1; i < n; i++ {
		res[i] = res[i-1] * nums[i-1]
	}
	// 后缀乘积，上三角，跳过nums[-1]
	tmp := 1
	for i := n - 2; i > -1; i-- {
		tmp *= nums[i+1]
		res[i] *= tmp
	}
	return res
}
